src/solc_0.8/bundleSandSale/PolygonBundleSandSale.sol

```diff
@@ -0,0 +1,281 @@
+        uint256 numPacks,
+        address to
+    ) external {
+        require(saleId > 0, "invalid saleId");
+        uint256 saleIndex = saleId - 1;
+        uint256 numPacksLeft = sales[saleIndex].numPacksLeft;
+        require(numPacksLeft >= numPacks, "not enough packs on sale");
+        sales[saleIndex].numPacksLeft = numPacksLeft - numPacks;
+
+        uint256 usdRequired = numPacks * sales[saleIndex].priceUSD;
+        require(_dai.transferFrom(msg.sender, _receivingWallet, usdRequired), "failed to transfer dai");
+        _transferPack(saleIndex, numPacks, to);
+
+        emit BundleSold(saleId, msg.sender, numPacks, address(_dai), usdRequired);
+    }
+
+    /**
+     * @notice get a specific sale information
+     * @param saleId id of the bundle
+     * @return priceUSD price in USD
+     * @return numPacksLeft number of packs left
+     */
+    function getSaleInfo(uint256 saleId) external view returns (uint256 priceUSD, uint256 numPacksLeft) {
+        require(saleId > 0, "invalid saleId");
+        uint256 saleIndex = saleId - 1;
+        priceUSD = sales[saleIndex].priceUSD;
+        numPacksLeft = sales[saleIndex].numPacksLeft;
+    }
+
+    /**
+     * @notice Remove a sale returning everything to some address
+     * @param saleId id of the bundle
+     * @param to The address that will receive the SAND
+     */
+    function withdrawSale(uint256 saleId, address to) external onlyAdmin {
+        require(saleId > 0, "invalid saleId");
+        uint256 saleIndex = saleId - 1;
+        uint256 numPacksLeft = sales[saleIndex].numPacksLeft;
+        sales[saleIndex].numPacksLeft = 0;
+
+        uint256[] memory ids = sales[saleIndex].ids;
+        uint256[] memory amounts = sales[saleIndex].amounts;
+        uint256 numIds = ids.length;
+        for (uint256 i = 0; i < numIds; i++) {
+            amounts[i] = amounts[i] * numPacksLeft;
+        }
+        require(
+            _sand.transferFrom(address(this), to, numPacksLeft * sales[saleIndex].sandAmount),
+            "transfer fo Sand failed"
+        );
+        _asset.safeBatchTransferFrom(address(this), to, ids, amounts, "");
+    }
+
+    /**
+     * @notice IERC1155TokenReceiver callback, creates a new Sale
+     */
+    function onERC1155Received(
+        address operator,
+        address from,
+        uint256 id,
+        uint256 value,
+        bytes calldata data
+    ) external override returns (bytes4) {
+        require(address(_asset) == msg.sender, "only accept asset as sender");
+        require(from == operator, "only self executed transfer allowed");
+        require(value > 0, "no Asset transfered");
+        require(data.length > 0, "data need to contains the sale data");
+
+        (uint256 numPacks, uint256 sandAmountPerPack, uint256 priceUSDPerPack) =
+            abi.decode(data, (uint256, uint256, uint256));
+
+        uint256 amount = value / numPacks;
+        require(amount * numPacks == value, "invalid amounts, not divisible by numPacks");
```

Comment:
Check what was the desired behavior in the case of a NFT  (value == 1)?

Reply:
When you create a NFT (only one of a kind) you cannot create packs, is kind of expected but is a strange behavior so may be we must add a specific comment or something or this comment can be just ignored.

Reply:
Well it's what 1155 is about, you can have multiple tokens of the same kind.
`value` is the number of tokens of the kind `id`

Reply:
When you transfer x assets, you except this number to be divisible by the number of packs. If you transfer only one asset, i guess you can only have one pack.
It's worth adding a comment to the method indeed